src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs

```diff
@@ -196,6 +196,12 @@ internal static class CSharpDiagnosticFilter
             {
                 // 4. Global analyzer config level
                 isSpecified = true;
+
+                // '/warnaserror' should promote warnings configured in global analyzer config to error.
+                if (report == ReportDiagnostic.Warn && generalDiagnosticOption == ReportDiagnostic.Error)
```

Comment:
Why do we only consider `generalDiagnosticOption` in this particular `if` block? Why not the others essentially?

Reply:
This is due to the fact that global configs are lazily parsed, while rulesets are parsed upfront by the command line parser.

1. Rulesets are parsed before parsing `/warnaserror` command line arguments: http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/CommandLine/CSharpCommandLineParser.cs,158
2. When parsing `/warnaserror` on command line, options from rulesets are directly bumped to errors in place: http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/CommandLine/CSharpCommandLineParser.cs,828

Reply:
By the way, we can consider implementing a similar fix for settings in editorconfig files, i.e. if you specify `dotnet_diagnostic.RuleId.severity = warning` in an editorconfig and execute with `/warnaserror`, then the diagnostic gets promoted to error - that does seem more sensible behavior as `/warnaserror` should ideally imply one does not see any warnings, but only errors (unless of course users have a `/warnaserror-:RuleId` for specific rules). Customers are also requesting the fix to editorconfig behavior in https://github.com/dotnet/roslyn/issues/43051, but there is potential of that being a more visible breaking change for existing users as editorconfig support in compiler has been shipped for a while. We have 2 options:

1. Avoid breaking change for editorconfig settings, and ask users to instead switch to global config if they desire `/warnaserror` to bump user configured warnings to errors. The approach taken currently in this PR.
2. Make the more correct fix of doing this change for both editorconfig and global config settings, at the risk of breaking some customers who have configured warnings in editorconfig files and run with `/warnaserror`. This would be a simple change to make in this PR if we are comfortable with it.

Reply:
Where do we document what the precedence level is between all of the various options here? I was a bit surprised at first glance to see that the test verifies that the command line essentially overrides the config files. Likely my intuition is just off here. But wanted to find a full list of "A wins over B" so I could consider what other tests we may need here.

Reply:
We have https://github.com/dotnet/roslyn/issues/47154 tracking adding this documentation.

Reply:
@jaredpar @chsienki I plan to add the precedence level documentation tracked in https://github.com/dotnet/roslyn/issues/47154 to docs.microsoft.com later today.

However, the open question in my above comment https://github.com/dotnet/roslyn/pull/47710#discussion_r490676970 still needs to be answered to define this order wrt to command line options: do we believe that command line options (/nowarn and /warnaserror) should always override diagnostic settings from any of the other sources (editorconfig, global config and rulesets)? I think that should be the case, as shown by the scenarios in https://github.com/dotnet/roslyn/issues/43051.

This PR ensures these semantics for global configs. I can create a follow-up PR for ensuring the same semantics for editorconfigs, if we feel it is a reasonable behavior change.

Reply:
@jaredpar Do you feel scheduling a meeting for discussing this would be preferrable? Also, should this PR be retargeted to 16.8 release branch?

Reply:
Sorry I just lost track of this.